LibraBFT/Impl/Consensus/SafetyRules/Properties/SafetyRules.agda

```diff
@@ -0,0 +1,717 @@
+    epoch = voteProposal ^∙ vpBlock ∙ bEpoch
+    round = voteProposal ^∙ vpBlock ∙ bRound
+
+    contract
+      : ∀ pre
+        → LBFT-weakestPre (constructAndSignVoteM maybeSignedVoteProposal) (Contract pre epoch round) pre
+    contract pre .unit refl nothing vs≡ ._ refl .unit refl =
+      record { noOuts = refl ; inv = record { es≡₁ = refl ; es≡₂ = refl } ; resultCorrect = record { lv≡ = refl ; lvr≤ = ≤-refl } }
+    contract pre .unit refl (just validatorSigner) vs≡ =
+      RWST-⇒
+        (Contract pre epoch round) Post pf
+        (constructAndSignVoteM-continue0 voteProposal validatorSigner) unit pre
+        (continue0.contract voteProposal validatorSigner pre)
+      where
+      Post : LBFT-Post (Either FakeErr Vote)
+      Post x post outs =
+        RWST-weakestPre-bindPost unit
+          (λ r → logInfo >> pure r) (Contract pre epoch round)
+          x post (LogInfo fakeInfo ∷ outs)
+
+      convNoOuts : ∀ {outs} → NoMsgOuts outs  → NoMsgOuts (LogInfo fakeInfo ∷ outs ++ LogInfo fakeInfo ∷ [])
+      convNoOuts{outs} pf rewrite List-filter-++ isOutputMsg? outs (LogInfo fakeInfo ∷ []) | pf = refl
+
+      pf : ∀ r st outs → Contract pre epoch round r st outs → Post r st outs
+      pf r st outs record { noOuts = noOuts ; inv = inv ; resultCorrect = resultCorrect } .r refl .unit refl =
+        record { noOuts = convNoOuts{outs} noOuts
+               ; inv = inv ; resultCorrect = resultCorrect }
+
+private
+  module Tutorial
+    (voteProposal : VoteProposal) (validatorSigner : ValidatorSigner)
+    (proposedBlock : Block)       (safetyData : SafetyData) where
+
+    -- After some experience with these proofs, it (allegedly)
+    -- becomes fairly straightforward to let Agda do a lot of the
+    -- work, and unfold the proof as we go.  However, it is
+    -- important to understand what's going on under the hood to be
+    -- able to reliably do this.  For the proof below, we do it in
+    -- excruciating detail "by hand" in comments as an example to
+    -- help ourselves understand.
+
+    -- For this example, we will prove that `step₃` of (and old version of)
+    -- `constructAndSignVoteM-continue2` produces no output.
+
+    Contract : LBFT-Post (Either FakeErr Vote)
+    Contract x post outs = outs ≡ []
+
+    step₃ : SafetyData → VoteData → Author → LedgerInfo → LBFT (Either FakeErr Vote)
+    step₃ safetyData1 voteData author ledgerInfo = do
+      let signature = ValidatorSigner.sign validatorSigner ledgerInfo
+          vote      = Vote.newWithSignature voteData author ledgerInfo signature
+      lSafetyData ∙= (safetyData1 & sdLastVote ?~ vote)
+      ok vote
+
+    step₃-contract
+      : ∀ pre safetyData1 voteData author ledgerInfo
+        → LBFT-weakestPre
+            (step₃ safetyData1 voteData author ledgerInfo)
+            Contract pre
+    step₃-contract pre safetyData1 voteData author ledgerInfo
+    {-
+    The proof can be as simple as this:
+
+       = λ _ _ _ _ _ _ → refl
```

Comment:
Making a note to myself that the essay still needs fixing.

Reply:
More generally, I have not reviewed all of the proofs in details (however, Agda has :smile:).  I don't want to hold up the merge to do so.  I'd like to go over especially the proofs labeled as "tutorial" and see how far they are from being understandable by mortals such as myself.  Then I can make or suggest changes later if needed.

Reply:
@mark-moir The fix I had in mind for the essay was only to remove two underscores. I am hoping to keep the proof in the private module `Tutorial` completely separate from the actual proof we use, because these proofs are going to undergo changes as more properties are tracked. There is certainly value in breaking up the larger proofs (that will be used to prove e.g. the votes once property) into smaller pieces, but I think the extended essay should be for something stand-alone since a considerable amount of effort goes into updating that essay if the model or property is modified.